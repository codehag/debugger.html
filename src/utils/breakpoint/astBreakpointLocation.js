import { getSymbols } from "../parser";
import { containsPosition } from "../parser/utils/helpers";

export async function getASTLocation(source, location) {
  const _source = source.toJS ? source.toJS() : source;
  const symbols = await getSymbols(_source);
  const functions = symbols.functions;

  const scope = functions.find(node => {
    return containsPosition(node.location, location);
  });
  if (scope) {
    const line = location.line - scope.location.start.line;
    const column = location.column;
    return { scope, offset: { line, column } };
  }
  return { scope: { name: undefined }, offset: location };
}

export async function findScopeByName(source, name) {
  const _source = source.toJS ? source.toJS() : source;
  const symbols = await getSymbols(_source);
  const functions = symbols.functions;

  return functions.find(node => {
    return node.name === name;
  });
}
