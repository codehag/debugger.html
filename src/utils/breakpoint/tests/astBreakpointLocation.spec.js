import { getASTLocation } from "../astBreakpointLocation.js";
import { getSource } from "../../parser/tests/helpers";

describe("ast", () => {
  describe("valid location", () => {
    it("returns the scope and offset", async () => {
      const source = getSource("math");
      // mock out immutable.js
      source.toJS = () => source;

      const location = { line: 6, column: 0 };
      const astLocation = await getASTLocation(source, location);
      const offset = astLocation.scope.location.start;
      expect(astLocation.scope.name).toBe("math");
      expect(astLocation.offset.line).toBe(location.line - offset.line);
      expect(astLocation.offset.column).toBe(location.column);
      expect(astLocation).toMatchSnapshot();
    });
  });
  describe("invalid location", () => {
    it("returns the scope and offset", async () => {
      const source = getSource("math");
      // mock out immutable.js
      source.toJS = () => source;

      const location = { line: 10, column: 0 };
      const astLocation = await getASTLocation(source, location);
      expect(astLocation.scope.name).toBe(undefined);
      expect(astLocation.offset.line).toBe(location.line);
      expect(astLocation.offset.column).toBe(location.column);
      expect(astLocation).toMatchSnapshot();
    });
  });
});
