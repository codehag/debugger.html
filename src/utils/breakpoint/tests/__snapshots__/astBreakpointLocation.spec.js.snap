// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast invalid location returns the scope and offset 1`] = `
Object {
  "offset": Object {
    "column": 0,
    "line": 10,
  },
  "scope": Object {
    "name": undefined,
  },
}
`;

exports[`ast valid location returns the scope and offset 1`] = `
Object {
  "offset": Object {
    "column": 0,
    "line": 5,
  },
  "scope": Object {
    "identifier": Object {
      "end": 13,
      "loc": Object {
        "end": Object {
          "column": 13,
          "line": 1,
        },
        "identifierName": "math",
        "start": Object {
          "column": 9,
          "line": 1,
        },
      },
      "name": "math",
      "start": 9,
      "type": "Identifier",
    },
    "location": Object {
      "end": Object {
        "column": 1,
        "line": 8,
      },
      "start": Object {
        "column": 0,
        "line": 1,
      },
    },
    "name": "math",
    "parameterNames": Array [
      "n",
    ],
  },
}
`;
